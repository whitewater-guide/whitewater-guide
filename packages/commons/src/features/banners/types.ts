import { NamedNode, Node } from '../../core';
import { Group } from '../groups';
import { Region } from '../regions';
import { Connection } from '../types';

export enum BannerPlacement {
  MOBILE_SECTION_DESCRIPTION = 'mobileSectionDescription',
  MOBILE_SECTION_ROW = 'mobileSectionRow',
  MOBILE_SECTION_MEDIA = 'mobileSectionMedia',
  MOBILE_REGION_DESCRIPTION = 'mobileRegionDescription',
}

export enum BannerKind {
  Image = 'Image',
  WebView = 'WebView',
}

export interface BannerSource {
  kind: BannerKind;
  ratio: number | null;
  src: string | null;
}

export interface Banner extends NamedNode {
  enabled: boolean;
  slug: string;
  priority: number;
  placement: BannerPlacement;
  source: BannerSource;
  link: string | null;
  extras: any;
  // --- connections
  regions?: Connection<Region>;
  groups?: Connection<Group>;
  // ---- transient
  deleted?: boolean;
}

export interface BannerInput {
  id: string | null;
  slug: string;
  name: string;
  enabled: boolean;
  priority: number;
  placement: BannerPlacement;
  source: BannerSource;
  link: string | null;
  extras: any;
  // --- connections
  regions: Node[];
  groups: Node[];
}

export function isBanner(node: Node): node is Banner {
  return node.__typename === 'Banner';
}
