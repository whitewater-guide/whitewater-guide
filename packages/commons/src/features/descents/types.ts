import { Timestamped, PageInfo } from '../../apollo';
import { Section } from '../sections';
import { User } from '../users';

export interface DescentLevel {
  __typename?: 'Level';
  unit?: string | null;
  value?: number | null;
}

export interface DescentLevelInput {
  unit?: string | null;
  value: number;
}

export interface Descent extends Timestamped {
  __typename?: 'Descent';
  id: string;
  user: User;
  section: Section;
  startedAt: string;
  duration?: number | null;
  level?: DescentLevel | null;
  comment?: string | null;
  public?: boolean | null;
}

export interface DescentEdge<TCursor = any> {
  __typename?: 'DescentEdge';
  node: Descent;
  cursor: TCursor;
}

export interface DescentInput {
  id?: string | null;

  sectionId: string;
  startedAt: string;
  duration?: number | null;
  level?: DescentLevelInput | null;
  comment?: string | null;
  public?: boolean | null;
}

export interface DescentsConnection<TCursor = any> {
  __typename?: 'DescentsConnection';
  edges: DescentEdge[];
  pageInfo: PageInfo<TCursor>;
}

export interface DescentsFilter {
  startDate?: string | null;
  endDate?: string | null;
  difficulty?: [number, number] | null;
  userId?: string | null;
  sectionId?: string | null;
  sectionName?: string | null;
}
