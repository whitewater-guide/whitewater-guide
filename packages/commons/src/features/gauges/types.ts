import { NamedNode, Node, Timestamped } from '../../core';
import { HarvestStatus, Measurement } from '../measurements';
import { Point, PointInput } from '../points';
import { Source } from '../sources';

export interface Gauge extends NamedNode, Timestamped {
  code: string;
  levelUnit: string | null;
  flowUnit: string | null;
  requestParams: any;
  cron: string | null;
  url: string | null;
  enabled: boolean;
  location: Point;
  source: Source;
  // measurements?: Measurement[];
  lastMeasurement: Measurement | null;
  status: HarvestStatus | null;
}

export interface GaugeInput {
  id: string | null;
  name: string;
  source: Node;
  location: PointInput | null;
  code: string;
  levelUnit: string | null;
  flowUnit: string | null;
  requestParams: any;
  cron: string | null;
  url: string | null;
}
