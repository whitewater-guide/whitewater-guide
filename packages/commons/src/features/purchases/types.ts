export enum PurchasePlatform {
  ios = 'ios',
  android = 'android',
  boomstarter = 'boomstarter',
}

export interface PurchaseInput {
  platform: PurchasePlatform;
  transactionId: string; // Promocode for boomstarter
  transactionDate?: Date; // or is it?
  productId: string;
  receipt?: string; // Can be undefined for boomstarter
  extra?: any;
}

export interface BoomPromoInfo {
  id: string;
  code: string;
  groupName: string | null;
  groupSku: string | null;
  redeemed: boolean;
}
