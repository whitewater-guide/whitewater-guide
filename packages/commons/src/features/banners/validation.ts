import isNumber from 'lodash/isNumber';
import { customStruct } from '../../utils/validation';
import { BannerKind, BannerPlacement } from './types';

const struct = customStruct({
  ratio: (value: any) => {
    if (!isNumber(value)) {
      return 'Ratio must be a number';
    }
    if (value > 10 || value < 3) {
      return 'Ratio must be a number between 10 and 3';
    }
    return true;
  },
});

const BannerSourceStruct = struct.object({
  kind: struct.enum(Object.values(BannerKind)),
  ratio: 'ratio|null',
  src: 'nonEmptyString',
});

export const BannerInputStruct = struct.object({
  id: 'uuid|null',
  slug: 'slug',
  name: 'nonEmptyString',
  priority: 'integer',
  enabled: 'boolean',
  placement: struct.enum(Object.values(BannerPlacement)),
  source: BannerSourceStruct,
  link: 'https|null',
  extras: 'object|null',
  regions: ['node'],
  groups: ['node'],
});
