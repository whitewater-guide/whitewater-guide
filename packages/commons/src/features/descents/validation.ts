import { yupTypes } from '@whitewater-guide/validation';
import * as yup from 'yup';
import { DescentLevelInput, DescentInput } from './types';

export const DescentLevelInputSchema = yup
  .object<DescentLevelInput>({
    unit: yup
      .string()
      .max(20)
      .notRequired()
      .nullable(),
    value: yup.number().required(),
  })
  .strict(true)
  .noUnknown();

export const DescentInputSchema = yup
  .object<DescentInput>({
    id: yupTypes
      .uuid()
      .notRequired()
      .nullable(),
    sectionId: yupTypes
      .uuid()
      .required()
      .nullable(false),
    startedAt: yup.date().required() as any,
    duration: yup
      .number()
      .integer()
      .notRequired()
      .nullable(),
    level: DescentLevelInputSchema.clone()
      .notRequired()
      .nullable(),
    comment: yup
      .string()
      .notRequired()
      .nullable(),
    public: yup
      .bool()
      .notRequired()
      .nullable(),
  })
  .strict(true)
  .noUnknown();
