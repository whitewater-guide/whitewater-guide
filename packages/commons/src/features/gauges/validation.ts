import { baseStruct } from '../../utils/validation';
import { PointInputStruct } from '../points';

const GaugeInputFields = {
  id: baseStruct.union(['uuid', 'null']),
  name: 'nonEmptyString',
  code: 'nonEmptyVarchar',
  levelUnit: 'nonEmptyVarchar|null',
  flowUnit: 'nonEmptyVarchar|null',
  location: baseStruct.union([PointInputStruct, 'null']),
  requestParams: 'object|null',
  cron: baseStruct.union(['cron', 'null']),
  url: baseStruct.union(['url', 'null', baseStruct.literal('')]),
  source: 'node',
};

export const GaugeInputStruct = baseStruct.object(GaugeInputFields);

export const GaugeFormStruct = baseStruct.object({
  ...GaugeInputFields,
  requestParams: baseStruct.union(['jsonString', 'null']),
});
