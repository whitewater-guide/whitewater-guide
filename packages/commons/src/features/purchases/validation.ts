import { yupTypes } from '@whitewater-guide/validation';
import * as yup from 'yup';

import { PurchaseInput, PurchasePlatform } from './types';

export const PurchaseInputSchema = yup
  .object<PurchaseInput>({
    platform: yup.mixed().defined().oneOf(Object.values(PurchasePlatform)),
    transactionId: yupTypes.nonEmptyString().defined().nullable(false),
    transactionDate: yup.date().notRequired().nullable(),
    productId: yupTypes.nonEmptyString().defined().nullable(false),
    receipt: yup.string().notRequired().nullable(),
    extra: yup.object().notRequired().nullable(),
  })
  .strict(true)
  .noUnknown();
