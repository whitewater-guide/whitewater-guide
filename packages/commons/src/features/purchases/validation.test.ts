import { createValidator } from '../../utils/validation';
import { PurchaseInput, PurchasePlatform } from './types';
import { PurchaseInputStruct } from './validation';

const validator = createValidator(PurchaseInputStruct);

type TestValue = [string, PurchaseInput];

const correct: PurchaseInput = {
  platform: PurchasePlatform.android,
  productId: 'region.altai',
  transactionId: '123',
  transactionDate: new Date(),
  receipt: '456',
  extra: { manual: true },
};

const correctValues: TestValue[] = [
  [
    'full value',
    correct,
  ],
  [
    'null value',
    {
      ...correct,
      transactionDate: null,
      receipt: null,
      extra: null,
    },
  ],
  [
    'empty value',
    {
      platform: PurchasePlatform.android,
      productId: 'region.altai',
      transactionId: '123',
    },
  ],
];

const incorrectValues: TestValue[] = [
  [
    'bad platform',
    { ...correct, platform: 'windows' } as any,
  ],
  [
    'bad product',
    { ...correct, productId: '' },
  ],
];

it.each(correctValues)('should be valid for %s', (_, value) => {
  expect(validator(value)).toBeNull();
});

it.each(incorrectValues)('should be invalid for %s', (_, value) => {
  expect(validator(value)).not.toBeNull();
  expect(validator(value)).toMatchSnapshot();
});
