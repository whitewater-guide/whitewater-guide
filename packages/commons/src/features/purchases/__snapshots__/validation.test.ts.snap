// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be invalid for bad platform 1`] = `
Object {
  "platform": "Expected a value of type \`\\"ios\\" | \\"android\\" | \\"boomstarter\\"\` for \`platform\` but received \`\\"windows\\"\`.",
}
`;

exports[`should be invalid for bad product 1`] = `
Object {
  "productId": "Value must be non-empty string",
}
`;
